### ISSUE TITLE
Zone Node Name not unique in AirLoopHVAC:ReturnPlenum object - need to issue error

### ISSUE THREAD

**Issue by BobHenninger** [NREL/EnergyPlus#4407]
Helpdesk Ticket 9254

> On 14 Apr 2014, at 21:42, "EnergyPlus Support" <support@energyplus.helpserve.com> wrote:
 > 
 > Finally(!) figured out what was wrong with this one.
 > 
 > AirLoopHVAC:ReturnPlenum,
 > Office FCU Return Plenum,!- Name
 > Plenum zone, !- Zone Name
 > Office Zone Air Node, !- Zone Node Name
 > Plenum Return Air Outlet,!- Outlet Node Name
 > , !- Induced Air Outlet Node or NodeList Name
 > Office FCU Return Air Outlet; !- Inlet 1 Node Name
 > 
 > The zone node name is the same as the Office zone air node name. The zone node name here should be something unique (not used anywhere else). EnergyPlus should throw an error if this node name is used anywhere else. If you change "Office Zone Air Node" in this object to "Plenum Zone Air Node" it will work as expected.
 > 
 > Mike
 > 
 > 
 > > But all following time steps is 24, however plenum zone temperature is
 > > higher each time.
 > > 
 > > From: EnergyPlus Support [mailto:support@energyplus.helpserve.com] 
 > > Sent: Friday, March 28, 2014 11:41 PM
 > > To: artem.zhukov@hotmail.com
 > > Subject: [EnergyPlus #9254]: Plenum zone return path 5-35
 > > 
 > > Plenum node outlet temperatures are always lagged by one zone
 > > timestep. So, if you report at the timestep frequency, if the plenum
 > > zone temperature is 24C at timestep n, then the plenum outlet node
 > > temperature will be 24C at timestep n+1. The plenum inlet temperature
 > > (zone return temperature) is added to the plenum zone heat balance at
 > > the end of the timestep, so the reported plenum temperature will
 > > include that effect, but the plenum outlet node will not change until
 > > the following timestep.
 > > Mike
 > > 
 > > > I have a plenum zone which has at the given timestamp the
 > > temperature
 > > > 24.51C, but I can't understand why the plenum return outlet is still
 > > > 24C?
 > > > Moreover if I add in lighting return fraction 0.3 the zone return
 > > > outlet is
 > > > increasing its temperature to 24.24C, but still plenum return outlet
 > > > is 24C
 > > > and all further nodes. How is it possible?
 > > > 
 > > > BR,
 > > > 
 > > > Artem Zhukov

User files
cm_fCube_f-var_w-var-VAV_luminar.idf

**Comment by rglazer2** [comment]
@mjwitte What is the best way to check if the node name is unique. I tried using
`CheckUniqueNodes("Return Plenum Zone Node Names", "NodeName", UniqueNodeError, ZoneRetPlenCond(ZonePlenumNum).ZoneNodeName);`
around ZonePlenum.cc:370 but that doesn't seem to find the other uses of "Office Zone Air Node".

**Comment by mjwitte** [comment]
@rglazer2 Looking more at other uses of CheckUniqueNodes, it is intended for a more limited scope so it will not do the trick here.  Because the return plenum node name (and supply plenum node names, and zone node names as well) could be re-used in another object, the validity check needs to be done after all the input has been processed.  These types of checks are done in [BranchNodeConnections::CheckNodeConnections](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/BranchNodeConnections.cc#L243).

Looking through the various validity checks, I don't see any check related to nodes of type NodeConnectionType_ZoneNode.  This new check can be described as "Check 11 - zone nodes may not be used as anything else except as a setpoint, sensor or actuator node".  So, checking the list of all nodes of type=NodeConnectionType_ZoneNode, there should not be another occurrence of the same node in the NodeConnections list with a type other than NodeConnectionType_Sensor, NodeConnectionType_Actuator, or NodeConnectionType_Setpoint.

As I look at this, I see that all of these checks are string comparisons and they don't need to be.  Opening a new issue #5138 to change that.

**Comment by rraustad** [comment]
Added new check in BranchNodeConnections.

```
** Severe  ** Node Connection Error, Node="OFFICE ZONE AIR NODE", Zone node name used in AirloopHVAC:ReturnPlenum is not unique.
**   ~~~   ** Reference Object=AIRLOOPHVAC:RETURNPLENUM, Name=OFFICE FCU RETURN PLENUM
************* There was 1 node connection error noted.
**  Fatal  ** Previous Conditions cause program termination.


// Check 11 -- AirLoopHVAC:ReturnPlenum zone node name must be unique
for( Loop1 = 1; Loop1 <= NumOfNodeConnections; ++Loop1 ) {
    if( NodeConnections( Loop1 ).ConnectionType != ValidConnectionTypes( NodeConnectionType_ZoneNode ) ) continue;
    if( NodeConnections( Loop1 ).ObjectType != "AIRLOOPHVAC:RETURNPLENUM" ) continue;
    MatchedAtLeastOne = false;
    for( Loop2 = 1; Loop2 <= NumOfNodeConnections; ++Loop2 ) {
        if( Loop1 == Loop2 ) continue;
        if( NodeConnections( Loop1 ).NodeNumber != NodeConnections( Loop2 ).NodeNumber ) continue;
        if( NodeConnections( Loop2 ).ConnectionType == ValidConnectionTypes( NodeConnectionType_ZoneNode ) ) {
            MatchedAtLeastOne = true;
            continue;
        }
    }
    if( MatchedAtLeastOne ) {
        ShowSevereError( "Node Connection Error, Node=\"" + NodeConnections( Loop1 ).NodeName + "\", Zone node name used in AirloopHVAC:ReturnPlenum is not unique." );
        ShowContinueError( "Reference Object=" + NodeConnections( Loop1 ).ObjectType + ", Name=" + NodeConnections( Loop1 ).ObjectName );
        ++ErrorCounter;
        ErrorsFound = true;
    }
}
```

**Comment by mjwitte** [comment]
@rraustad There's already a fix waiting for this one #5150 - but it breaks the secondary zone condenser file - see #5174.  @Myoldmopar was going to fix that, then I took it over, but I haven't done any code yet.  Let me know if you want to take that one.

**Comment by rraustad** [comment]
Outstanding, I just spent most of the day working out the unit test for this 10 line fix. #5150 logic differs from mine since I specifically look for a return plenum and only then fatal if found more than once. I'm pushing it up to see if any example files fail.

**Comment by rraustad** [comment]
This specific issue addressed at https://github.com/NREL/EnergyPlus/pull/5338

**Comment by Myoldmopar** [comment]
Closed via #5338
